Central-Server/API/internal/threads.py

```diff
@@ -0,0 +1,398 @@
+        self.queue.insert(0, job) # Insert job into the front of the queue
+
+        # Update the status of the job in the database.
+        conn = database.connect()
+        cursor = conn.cursor()
+        cursor.execute("UPDATE hilsim_runs set run_status = %s, run_start = now() where run_id = %s",
+                        (jobs.JobStatus.QUEUED.value, job.data["job_data"]["job_id"]))
+        cursor.close()
+
+    def get_recent_queue(self) -> List: # List of what? TODO: check
+        """Gets all jobs in the queue that happened after the last time the queue was checked by the server"""
+        conn = database.connect()
+        cursor = conn.cursor()
+        delta_t = (datetime.datetime.now() - self.last_time).total_seconds()
+        db_query = "SELECT * from hilsim_runs where hilsim_runs.run_status = %s and hilsim_runs.submitted_time > %s ORDER BY submitted_time"
+        cursor.execute(db_query, (jobs.JobStatus.QUEUED.value, self.last_time))
+        queued_jobs = cursor.fetchall()
+        if len(queued_jobs) > 0:
+            # Then get the last one
+            struct = database.convert_database_tuple(cursor, queued_jobs[-1])
+            self.last_time = struct.submitted_time
+        return database.convert_database_list(cursor, queued_jobs)
+
+    def some_thread_active(self) -> bool:
+        """Returns TRUE if there exists a thread capable of taking a job. Otherwise FALSE"""
+        for t in self.threads:
+            if t.is_alive():
+                if not t.can_take_job():
+                    return True
+            else:
+                del t
+        return False
+
+    def create_threads(self):
+        """Create all threads queued for initialization"""
+        global GLOBAL_BOARD_ID
+        while len(self.spin_up_queue) > 0:
+            t = self.spin_up_queue.pop(0)
+            thr = BoardThread(t.thread_name, t.thread_name, GLOBAL_BOARD_ID, t.communicaton_channel, self.pop_killed_job_back_to_queue)
+            GLOBAL_BOARD_ID += 1 # Assign a new board ID and increment the global id
+            thr.start()
+            self.threads.append(thr)
+
+    def add_job(self, config):
+        """Adds a job to the queue"""
+        print(f"added job {config} to queues", flush=True)
+        self.queue.append(config)
+    
+    def add_thread(self, thr):
+        """Adds a thread to be spun up"""
+        self.spin_up_queue.append(thr)
+    
+    def terminate_all(self):
+        """Destroy all threads and manager thread"""
+        for t in self.threads:
+            if t.is_alive():
+                t.terminate()
+                print(f"terminated thread {t.thread_id}", flush=True)
+        
+        self.running = False
+        print(f"terminated manager", flush=True)
+        
+    def remove_dead_threads(self):
+        """Remove threads that are not 'alive'"""
+        self.threads = [thr for thr in self.threads if thr.is_alive()]
+    
+    def kill_thr(self, ident):
+        # TODO

```

Comment:
Do you keep track of all the things that are TODO? Maybe in your roadmap/trello. If they're already in the roadmap, add some more information so that you don't lose track of where things are and what you're supposed to do.

Reply:
I think TODO comments signify low hanging low priority fruit, so it is a bit overkill to put it in the roadmap and trello because maintaining them will take some time.
They already serve as a marker, so you just Ctrl-F the repo for any TODOs, if you want to feel like you contributed.

Reply:
Good idea, we'll add it to the roadmap

Reply:
@zyunlam im gonna push back on that a bit bc the point of a ticket isn't to signify large contributions that need to be made, it's to track work that needs to be done. if it's gonna add clutter to the trello board, i'd say that demands a need for separate trello boards or like a hotfix tag on those tickets. it'll also make it easier to contribute since ppl would know if others r alr working on that todo without having to comb through branches. overall streamlines the process

Reply:
@zyunlam im gonna push back on that a bit bc the point of a ticket isn't to signify large contributions that need to be made, it's to track work that needs to be done. if it's gonna add clutter to the trello board, i'd say that demands a need for separate trello boards or like a hotfix tag on those tickets. it'll also make it easier to contribute since ppl would know if others r alr working on that todo without having to comb through branches. overall streamlines the process

Reply:
I'm with Rithvik and Aidan on this one, though I feel like this should only be done for TODOs where actual code needs to be written -- some TODOs are marked as TODO to just.. check if something is true or whatever.

We'll discuss as a team next meeting.